# Credit Risk: Ensemble Learners Notebook

This notebook seeks to train and compare two different ensemble classifiers to predict loan risk. Additionally, to evaluate each model the SciKit-Learn and Imbalanced-Learn libraries will be utilized. 

---

## Import the Required Libraries & Dependencies

In [1]:
# Import Warnings 
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Import Libraries and Dependencies 
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
# Import SK-LEARN & IM-LEARN libraries 
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

---

## Read the CSV and Perform Basic Data Cleaning


In [4]:
# Load the data
file_path = Path('Resources/lending_data.csv')
df = pd.read_csv(file_path)

In [5]:
# Display the df
df.head()

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
0    10700.0          7.672       own            52800        0.431818   
1     8400.0          6.692       own            43600        0.311927   
2     9000.0          6.963      rent            46100        0.349241   
3    10700.0          7.664       own            52700        0.430740   
4    10800.0          7.698  mortgage            53000        0.433962   

   num_of_accounts  derogatory_marks  total_debt loan_status  
0                5                 1       22800    low_risk  
1                3                 0       13600    low_risk  
2                3                 0       16100    low_risk  
3                5                 1       22700    low_risk  
4                5                 1       23000    low_risk

### Brief Data Cleaning

**Important Section Note:** 

It has been pre-determined the given LendingClub data has previously undergone cleaning procedures. The following section iterates the cleaning process to present possible data cleaning methods if it was required. 

#### Data Cleaning Part 1: Remove Null Values 

In [6]:
# Identify null values in the data set
# df = df.isnull.sum()

# Remove null values from the data set
# df = df.dropna(inplace = True)

# Verfiy null values have been removed from the data set 
# df.isnull().sum()

#### Data Cleaning Part 2: Identify Column Errors

The following data cleaning section cotnains a hypothetical scenario that could be present in a dataset. 

**Hypothetical Scenario:**

The section column: "borrower_income", returns an object and contains '$'

In [7]:
# Identify column errors
 
# Check the data types
# df.dtypes

# Remove the '$' in the "borrower_income" column 
# df['borrower_income'] = df['borrower_income'].str.replace('$', '')

# Display the df
# df.head()

# The column 'borrower_income' returns an object, change it to a float 
# df['borrower_income'] = df['borrower_income'].astype('float')

# Check if the data types are correct 
# df.dtypes


#### Data Cleaning Part 3: Summary Statistics

In [8]:
# Summary statistics

# df[df['loan_status'] == "low_risk"].describe()
# df[df['loan_status'] == "high_risk"].describe()

---

## Split the Data into Training and Testing

Split the data into training and testing categories. 


In [9]:
# Create our features
X = df.copy()
X = X.drop(columns = 'loan_status')
X = pd.get_dummies(X)

# Create our target
y = df['loan_status']

In [10]:
# Display Summary Statistics 
X.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  homeowner_mortgage  \
count     77536.000000      77536.000000  77536.000000        77536.000000   
mean          3.826610          0.392308  19221.949804            0.497472   
std           1.904426          0.582086   8371.635077            0.499997   
min           0.000000          0.000000      0.000000            0.000000   
25%           3.000000          0.000000  14800.000000            0.000000   
50%           4.000000          0.000000  18100.000000            0.000000   
75%           4.000000          1.000000  21400.000000            1.000000   
max          16.000000          3.000000  75200.000000            1.000000   

       homeowner_own  homeowner_rent  
count   77536.000000    77536.000000  
mean        0.398911        0.103616  
std         0.489678        0.304764  
min         0.000000        0.000000  
25%         0.000000        0.000000  
50%         0.000000        0.000000  
75%         1.000000        0.000000  
max         1.000000        1.000000

In [11]:
# Check the balance of our target values
y.value_counts()

low_risk     75036
high_risk     2500
Name: loan_status, dtype: int64

### Split the X and y into X_train, X_test, y_train, y_test

In [12]:
# Import train_test_split from sklearn
from sklearn.model_selection import train_test_split

In [13]:
# Split the data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

# Shape X_train
X_train.shape

(58152, 10)

In [14]:
# View the count for y_train
Counter(y_train)

Counter({'low_risk': 56277, 'high_risk': 1875})

---

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Print the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier onely, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests


## Random Over Sampler

**Important Section Note:**

This section of the notebook contains the Random Over Sampler model form the orginal Ensembe Techniques starter code notebook. Despite the updated starter code notebook not including this step, it has been included to ensure there has been no confusion between both notebooks. 


In [15]:
# Import Random Over Sampler (ROS)
from imblearn.over_sampling import RandomOverSampler

In [16]:
# Train the Random Over Sampler (ROS)
ROS = RandomOverSampler(random_state = 1)

# Fit the model
X_resampled, y_resampled = ROS.fit_resample(X_train, y_train)

# View the count for y_train
Counter(y_resampled)

Counter({'low_risk': 56277, 'high_risk': 56277})

In [17]:
# Logistic Regression Import (LR)
from sklearn.linear_model import LogisticRegression

In [18]:
# Train the Logistic Regression model using the resampled data
LR = LogisticRegression(solver='lbfgs', random_state=1, max_iter=100)

# Fit the model
LR.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

### Balanced Accuracy Score

In [19]:
# Calculated the balanced accuracy score
y_pred = LR.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9947746894823818

### Confusion Matrix

In [20]:
# Display the confusion matrix (CM)
CM_ROS = confusion_matrix(y_test, y_pred)

# Create a confusion matrix df (CM_df)
CM_ROS_df = pd.DataFrame(CM_ROS, 
                         index = ["High Risk (Predicted Positive)", "Low RIsk (Predicted Negative)"], 
                         columns = ["Positive", "Negative"])
# Display the CM_df
display(CM_ROS_df)

Positive  Negative
High Risk (Predicted Positive)       622         3
Low RIsk (Predicted Negative)        106     18653

### Classification Report

In [21]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.85      1.00      0.99      0.92      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



## Balanced Random Forest Classifier

In [22]:
# Import balanced random forest classifier (BRF)
from imblearn.ensemble import BalancedRandomForestClassifier

In [23]:
# Resample the training data with the Balanced Random Forest Classifier
# Train the model using BRF 
BRF = BalancedRandomForestClassifier(n_estimators=100, random_state=1)

# Fit the model
BRF.fit(X_train, y_train)

BalancedRandomForestClassifier(random_state=1)

### Balanced Accuracy Score

In [24]:
# Calculated the balanced accuracy score
y_pred = BRF.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9936818806972654

### Confusion Matrix

In [25]:
# Display the confusion matrix (CM)
CM_BRF = confusion_matrix(y_test, y_pred)

# Create a confusion matrix df (CM_df)
CM_BRF_df = pd.DataFrame(CM_BRF, 
                         index = ["High Risk (Predicted Positive)", "Low RIsk (Predicted Negative)"], 
                         columns = ["Positive", "Negative"])
# Display the CM_df
display(CM_BRF_df)

Positive  Negative
High Risk (Predicted Positive)       622         3
Low RIsk (Predicted Negative)        147     18612

### Classification Report

In [26]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.81      1.00      0.99      0.89      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       0.99      0.99      1.00      0.99      0.99      0.99     19384



### Feature Importance

In [27]:
# List the features sorted in descending order by feature importance
feature_importance = BRF.feature_importances_
sorted(zip(BRF.feature_importances_, X.columns), reverse = True)

[(0.22759633831687995, 'borrower_income'),
 (0.1965574723230606, 'interest_rate'),
 (0.17417807060091195, 'debt_to_income'),
 (0.14944360428565667, 'total_debt'),
 (0.14229163194622926, 'loan_size'),
 (0.07859627283289125, 'num_of_accounts'),
 (0.02885770833789564, 'derogatory_marks'),
 (0.001219411955649641, 'homeowner_own'),
 (0.0008974701883394341, 'homeowner_mortgage'),
 (0.0003620192124855904, 'homeowner_rent')]

---

## Easy Ensemble Classifier

The Easy Ensemble Classifier section contains both the Adaptive Boosting classifier and the Easy Ensable classifier. The original starter_code notebook stated AdaBoost would be used in this section of classification however, the updated starter_code notebook indicated the use of the Esay Ensemble Classifier. Due to this reason the final submission includes both classification models. 

In [28]:
# Import the Easy Ensemble Classifier (EEC)
from imblearn.ensemble import EasyEnsembleClassifier

In [29]:
# Train the EasyEnsembleClassifier (EEC)
EEC = EasyEnsembleClassifier(n_estimators=100, random_state=1)

# Fit the model
EEC.fit(X_train, y_train)

EasyEnsembleClassifier(n_estimators=100, random_state=1)

### Balanced Accuracy Score

In [30]:
# Calculated the balanced accuracy score
y_pred = EEC.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9944548430086891

### Confusion Matrix

In [31]:
# Display the confusion matrix
# Display the confusion matrix (CM)
CM_EEC = confusion_matrix(y_test, y_pred)

# Create a confusion matrix df (CM_df)
CM_EEC_df = pd.DataFrame(CM_EEC, 
                         index = ["High Risk (Predicted Positive)", "Low RIsk (Predicted Negative)"], 
                         columns = ["Positive", "Negative"])
# Display the CM_df
display(CM_EEC_df)

Positive  Negative
High Risk (Predicted Positive)       622         3
Low RIsk (Predicted Negative)        118     18641

### Classification Report

In [32]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.84      1.00      0.99      0.91      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       0.99      0.99      1.00      0.99      0.99      0.99     19384



## AdaBoost

In [33]:
# Import AdaBoost classifier from SkLearn (AdaB)
from sklearn.ensemble import AdaBoostClassifier

In [34]:
# Train the AdaBoost classifier (AdaB)
AdaB = AdaBoostClassifier(n_estimators=100, random_state=1)

# Fit the model
AdaB.fit(X_resampled, y_resampled)

AdaBoostClassifier(n_estimators=100, random_state=1)

### Balanced Accuracy Score

In [35]:
# Calculated the balanced accuracy score
y_pred = AdaB.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9947746894823818

### Confusion Matrix

In [36]:
# Display the confusion matrix
# Display the confusion matrix (CM)
CM_AdaB = confusion_matrix(y_test, y_pred)

# Create a confusion matrix df (CM_df)
CM_AdaB_df = pd.DataFrame(CM_AdaB, 
                          index = ["High Risk (Predicted Positive)", "Low RIsk (Predicted Negative)"], 
                          columns = ["Positive", "Negative"])
# Display the CM_df
display(CM_AdaB_df)

Positive  Negative
High Risk (Predicted Positive)       622         3
Low RIsk (Predicted Negative)        106     18653

### Classification Report

In [37]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.85      1.00      0.99      0.92      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



---

## Conclusion 

1. Which model had the best balanced accuracy score?
2. Which model had the best recall score?
3. Which model had the best geometric mean score?
4. What are the top three features?


### Balanced Accuracy Score Values:

**Oversampling -**

| Oversampler | Balanced Accuracy Score |
| --- | --- | 
| Naive Random Oversampling | 0.9947746894823818 | 

    
**Ensemble Learners -**

| Ensemble Learner | Balanced Accuracy Score |
| --- | --- | 
| Balanced RandomForest Classifier | 0.9936818806972654 | 
| Easy Ensemble Classifier | 0.9944548430086891 | 
| Adaptive Boosting | 0.9947746894823818 | 


The models with the **best balanced accuracy score** are the Niave Random Oversampling classifier and the Adaptive Boosting esemble learner. 

### Recall Score Values:

**Oversampling -**

| Oversampler | High Risk | Low Risk |
| --- | --- | --- |
| Naive Random Oversampling | 1.00 | 0.99 |


**Ensemble Learners -**

| Ensemble Learner | High Risk | Low Risk |
| --- | --- | --- |
| Balanced RandomForest Classifier | 1.00 | 0.99 |
| Easy Ensemble Classifier | 1.00 | 0.99 |
| Adaptive Boosting | 1.00 | 0.99 |


All models have an equal **recall score**. 

### Geometric Mean Score Values: 

**Oversampling -**

| Oversampler | High Risk | Low Risk |
| --- | --- | --- |
| Naive Random Oversampling | 0.99 | 0.99 |


**Ensemble Learners -**

| Ensemble Learner | High Risk | Low Risk |
| --- | --- | --- |
| Balanced RandomForest Classifier | 0.99 | 0.99 |
| Easy Ensemble Classifier | 0.99 | 0.99 |
| Adaptive Boosting | 0.99 | 0.99 |


All models have an equal **geometric mean score**. 

### Top 3 Features: 

The top three features are: 

| Column Feature | Feature Score |
| --- | --- | 
| Borrower Income | 0.22759 | 
| Interest Rate | 0.19655 | 
| Debt to Income | 0.17417 | 


---